lib/burnside/flattened.h

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_BURNSIDE_FLATTENED_H_
+#define FORTRAN_BURNSIDE_FLATTENED_H_
+
+#include "mixin.h"
+#include "../parser/parse-tree.h"
+#include "llvm/ADT/BitVector.h"
+#include <cstdint>
+#include <list>
+#include <map>
+#include <set>
+#include <vector>
+
+namespace Fortran::burnside {
+
+struct AnalysisData;
+
+namespace flat {
+
+/// This is a flattened, linearized representation of the parse
+/// tree. It captures the executable specification of the input
+/// program. The flattened IR can be used to construct FIR.
+///
+/// [Coding style](https://llvm.org/docs/CodingStandards.html)
+
+using LabelRef = unsigned;
+constexpr LabelRef unspecifiedLabel{UINT_MAX};
+
+using Location = parser::CharBlock;
+struct LabelBuilder;
+
+// target for a control-flow edge
+struct LabelOp {
```

Comment:
I take it that these ops are different from the fir ops & only used for this flattening.  Is that right?

Reply:
That is correct.